File: Daily-Streak/Day-12_Minimum-Time-Visiting-All-Points/Solution.cpp
class Solution {
public:
    int minTimeToVisitAllPoints(vector<vector<int>>& points) {
        // ✅ OPTIMAL GREEDY APPROACH (ACTIVE)
        int time = 0;

        for (int i = 1; i < points.size(); i++) {
            int dx = abs(points[i][0] - points[i - 1][0]);
            int dy = abs(points[i][1] - points[i - 1][1]);

            // Diagonal moves cover both x and y together
            time += max(dx, dy);
        }

        return time;
    }

    /*
    ===========================
    🧠 ALTERNATE THINKING (COMMENTED)
    ===========================
    - Each diagonal move reduces both dx and dy by 1
    - Remaining distance is straight moves
    - Total time = max(dx, dy)

    ===========================
    ❌ SIMULATION APPROACH (COMMENTED)
    ===========================
    - Move step by step
    - Simulate diagonal/straight moves
    - Correct but unnecessary

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    #include <bits/stdc++.h>
    using namespace std;

    int main() {
        vector<vector<int>> points = {{1,1},{3,4},{-1,0}};
        Solution s;
        cout << s.minTimeToVisitAllPoints(points) << endl;
        return 0;
    }
    */
};

File: Daily-Streak/Day-12_Minimum-Time-Visiting-All-Points/Solution.py
class Solution:
    def minTimeToVisitAllPoints(self, points: List[List[int]]) -> int:
        # ✅ OPTIMAL GREEDY APPROACH (ACTIVE)
        time = 0

        for i in range(1, len(points)):
            dx = abs(points[i][0] - points[i - 1][0])
            dy = abs(points[i][1] - points[i - 1][1])

            # Diagonal movement minimizes time
            time += max(dx, dy)

        return time

        """
        ===========================
        🧠 INTUITION (COMMENTED)
        ===========================
        - Diagonal move = (x+1, y+1)
        - Use diagonals until one axis aligns
        - Remaining moves are straight

        ===========================
        ❌ SIMULATION IDEA (COMMENTED)
        ===========================
        - While current != target:
            move step by step
        - Correct but verbose

        ===========================
        🧪 STANDALONE VERSION (COMMENTED)
        ===========================
        def main():
            points = [[1,1],[3,4],[-1,0]]
            sol = Solution()
            print(sol.minTimeToVisitAllPoints(points))

        if __name__ == "__main__":
            main()
        """
